### PR TITLE
Fix ADSC race

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Fix: #28341

The `ADSC` fields should be set before running. So i split the `Run` method out.

### CODE REVIEW

**Review by howardjohn — APPROVED** [review]

File: pkg/istio-agent/local_xds_generator.go [hunk 1]
```diff
@@ -144,10 +144,12 @@ func (sa *Agent) startXDSGenerator(proxyConfig *meshconfig.ProxyConfig, secrets
 		// we are in 'envoy using proxy' mode.
 		return fmt.Errorf("adsc: %v", err)
 	}
-
 	ads.LocalCacheDir = savePath.Get()
 	ads.Store = sa.localXDSGenerator.xdsServer.MemoryConfigStore
 	ads.Registry = sa.localXDSGenerator.xdsServer.DiscoveryServer.MemRegistry
+	if err := ads.Run(); err != nil {
+		return fmt.Errorf("adsc: failed running %v", err)
```
Comment (shamsher31) [inline pkg/istio-agent/local_xds_generator.go:151]: nit:
```suggestion
		return fmt.Errorf("ADSC: failed running %v", err)
```

File: pkg/adsc/adsc_test.go [hunk 2]
```diff
@@ -131,7 +131,10 @@ func TestADSC_Run(t *testing.T) {
 				t.Errorf("Dial error: %v", err)
 				return
 			}
-
+			if err := tt.inAdsc.Run(); err != nil {
+				t.Errorf("ADSC run error: %v", err)
```
Comment (shamsher31) [inline pkg/adsc/adsc_test.go:135]: nit:
```suggestion
				t.Errorf("ADSC: failed running %v", err)
```

File: pilot/pkg/xds/lds_test.go [hunk 3]
```diff
@@ -406,7 +411,9 @@ func TestLDSEnvoyFilterWithWorkloadSelector(t *testing.T) {
 				t.Fatal(err)
 			}
 			defer adsResponse.Close()
-
+			if err := adsResponse.Run(); err != nil {
+				t.Fatal("Error running adsc ", err)
```
Comment (shamsher31) [inline pilot/pkg/xds/lds_test.go:415]: I think consistent error message will help to identify `ADSC` related issue in logs?
```suggestion
				t.Fatal("ADSC: failed running ", err)
```

File: istioctl/pkg/xds/client.go [hunk 4]
```diff
@@ -59,6 +59,10 @@ func GetXdsResponse(dr *xdsapi.DiscoveryRequest, opts *clioptions.CentralControl
 	if err != nil {
 		return nil, fmt.Errorf("could not dial: %w", err)
 	}
+	err = adscConn.Run()
+	if err != nil {
+		return nil, fmt.Errorf("adsc: failed running %v", err)
```
Comment (shamsher31) [inline istioctl/pkg/xds/client.go:64]:
```suggestion
		return nil, fmt.Errorf("ADSC: failed running %v", err)
```

File: pilot/pkg/networking/apigen/apigen_test.go [hunk 5]
```diff
@@ -82,6 +82,10 @@ func TestAPIGen(t *testing.T) {
 
 		configController := memory.NewController(store)
 		adscConn.Store = model.MakeIstioStore(configController)
+		err = adscConn.Run()
+		if err != nil {
+			t.Fatal("Error running adsc ", err)
```
Comment (shamsher31) [inline pilot/pkg/networking/apigen/apigen_test.go:87]:
```suggestion
			t.Fatal("ADSC: failed running ", err)
```

File: pilot/pkg/xds/ads_test.go [hunk 6]
```diff
@@ -131,6 +131,9 @@ func TestAgent(t *testing.T) {
 			t.Fatal("Failed to connect", err)
 		}
 		defer ldsr.Close()
+		if err := ldsr.Run(); err != nil {
+			t.Fatal("Error running adsc ", err)
```
Comment (shamsher31) [inline pilot/pkg/xds/ads_test.go:135]:
```suggestion
			t.Fatal("ADSC: failed running", err)
```

File: pilot/pkg/xds/eds_test.go [hunk 7]
```diff
@@ -351,6 +351,9 @@ func adsConnectAndWait(t *testing.T) *adsc.ADSC {
 	if err != nil {
 		t.Fatal("Error connecting ", err)
 	}
+	if err := adscConn.Run(); err != nil {
+		t.Fatal("Error running adsc ", err)
```
Comment (shamsher31) [inline pilot/pkg/xds/eds_test.go:355]:
```suggestion
		t.Fatal("ADSC: failed running ", err)
```

File: pilot/pkg/xds/fake.go [hunk 8]
```diff
@@ -319,6 +319,10 @@ func (f *FakeDiscoveryServer) Connect(p *model.Proxy, watch []string, wait []str
 	if err != nil {
 		f.t.Fatalf("Error connecting: %v", err)
 	}
+	if err := adscConn.Run(); err != nil {
+		f.t.Fatalf("Error running adsc: %v", err)
```
Comment (shamsher31) [inline pilot/pkg/xds/fake.go:323]:
```suggestion
		f.t.Fatalf("ADSC: failed running: %v", err)
```

File: pilot/pkg/xds/lds_test.go [hunk 9]
```diff
@@ -154,33 +154,35 @@ func TestLDSWithDefaultSidecar(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer adsResponse.Close()
-
-	adsResponse.Watch()
+	defer adsc.Close()
+	if err := adsc.Run(); err != nil {
+		t.Fatal("Error running adsc ", err)
```
Comment (shamsher31) [inline pilot/pkg/xds/lds_test.go:159]:
```suggestion
		t.Fatal("ADSC: failed running ", err)
```

File: pilot/pkg/xds/lds_test.go [hunk 10]
```diff
@@ -215,29 +217,30 @@ func TestLDSWithIngressGateway(t *testing.T) {
...
-
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer adsResponse.Close()
-
-	adsResponse.Watch()
+	defer adsc.Close()
+	if err := adsc.Run(); err != nil {
+		t.Fatal("Error running adsc ", err)
```
Comment (shamsher31) [inline pilot/pkg/xds/lds_test.go:225]:
```suggestion
		t.Fatal("ADSC: failed running ", err)
```

File: pilot/pkg/xds/lds_test.go [hunk 11]
```diff
@@ -303,11 +306,13 @@ func TestLDSWithSidecarForWorkloadWithoutService(t *testing.T) {
...
```
Comment (shamsher31) [inline pilot/pkg/xds/lds_test.go:311]:
```suggestion
		t.Fatal("ADSC: failed running ", err)
```

**Review by shamsher31 — APPROVED** [review]
Thanks.

**Pull request merged**